src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -35,15 +35,25 @@ private async Task AnalyzeDocumentForKindAsync(TextDocument document, AnalysisKi
         {
             try
             {
-                if (!AnalysisEnabled(document, _documentTrackingService))
+                if (!document.SupportsDiagnostics())
+                {
+                    return;
+                }
+
+                var isActiveDocument = _documentTrackingService.TryGetActiveDocument() == document.Id;
+                var isOpenDocument = document.IsOpen();
+                var isGeneratedRazorDocument = document.Services.GetService<DocumentPropertiesService>()?.DiagnosticsLspClientName != null;
+
+                // Only analyze open/active documents, unless it is a generated Razor document.
+                if (!isActiveDocument && !isOpenDocument && !isGeneratedRazorDocument)
                 {
-                    // to reduce allocations, here, we don't clear existing diagnostics since it is dealt by other entry point such as
-                    // DocumentReset or DocumentClosed.
                     return;
                 }
 
                 var stateSets = _stateManager.GetOrUpdateStateSets(document.Project);
                 var compilationWithAnalyzers = await GetOrCreateCompilationWithAnalyzersAsync(document.Project, stateSets, cancellationToken).ConfigureAwait(false);
+                var version = await GetDiagnosticVersionAsync(document.Project, cancellationToken).ConfigureAwait(false);
+                var backgroundAnalysisScope = SolutionCrawlerOptions.GetBackgroundAnalysisScope(document.Project);
```

Comment:
note to self: this feels unclear.  why would the BGAnalysisScope be affected by a particular project?  it's a global option the user can set.

Reply:
Project is just passed to get to the options and language: https://github.com/dotnet/roslyn/blob/7b6b57276d7a27198956beb3a86d7d0db34a0829/src/Workspaces/Core/Portable/SolutionCrawler/SolutionCrawlerOptions.cs#L73-L74